code/modules/station_goals/dna_vault.dm

```diff
@@ -0,0 +1,276 @@
+	var/list/occupied = list()
+	for(var/direct in list(EAST,WEST,SOUTHEAST,SOUTHWEST))
+		occupied += get_step(src,direct)
+	occupied += locate(x+1,y-2,z)
+	occupied += locate(x-1,y-2,z)
+
+	for(var/T in occupied)
+		var/obj/structure/filler/F = new(T)
+		F.parent = src
+		fillers += F
+
+	if(ticker.mode)
+		for(var/datum/station_goal/dna_vault/G in ticker.mode.station_goals)
+			animals_max = G.animal_count
+			plants_max = G.plant_count
+			dna_max = G.human_count
+			break
+
+/obj/machinery/dna_vault/Destroy()
+	for(var/V in fillers)
+		var/obj/structure/filler/filler = V
+		filler.parent = null
+		qdel(filler)
+	. = ..()
+	
+/obj/machinery/dna_vault/attack_hand(mob/user)
+	if(..())
+		return 1
+	ui_interact(user)
+
+/obj/machinery/dna_vault/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "dna_vault.tmpl", name, 350, 400)
+		ui.open()
+
+/obj/machinery/dna_vault/proc/roll_powers(mob/user)
+	if(user in power_lottery)
+		return
+	var/list/L = list()
+	var/list/possible_powers = list(VAULT_TOXIN,VAULT_NOBREATH,VAULT_FIREPROOF,VAULT_STUNTIME,VAULT_ARMOUR)
+	L += pick_n_take(possible_powers)
+	L += pick_n_take(possible_powers)
+	power_lottery[user] = L
+
+/obj/machinery/dna_vault/ui_data(mob/user, ui_key = "main", datum/topic_state/state = physical_state) //TODO Make it % bars maybe
+	var/list/data = list()
+	data["plants"] = plants.len
+	data["plants_max"] = plants_max
+	data["animals"] = animals.len
+	data["animals_max"] = animals_max
+	data["dna"] = dna.len
+	data["dna_max"] = dna_max
+	data["completed"] = completed
+	data["used"] = TRUE
+	data["choiceA"] = ""
+	data["choiceB"] = ""
+	if(user && completed)
+		var/list/L = power_lottery[user]
+		if(L && L.len)
+			data["used"] = FALSE
+			data["choiceA"] = L[1]
+			data["choiceB"] = L[2]
+	return data
+
+/obj/machinery/dna_vault/Topic(href, href_list)
+	if(..())
+		return 1
+
+	if(href_list["gene"])
+		upgrade(usr,href_list["choice"])
+		. = TRUE
+
+/obj/machinery/dna_vault/proc/check_goal()
+	if(plants.len >= plants_max && animals.len >= animals_max && dna.len >= dna_max)
+		completed = TRUE
+
+/obj/machinery/dna_vault/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/device/dna_probe))
+		var/obj/item/device/dna_probe/P = I
+		var/uploaded = 0
+		for(var/plant in P.plants)
+			if(!plants[plant])
+				uploaded++
+				plants[plant] = 1
+		for(var/animal in P.animals)
+			if(!animals[animal])
+				uploaded++
+				animals[animal] = 1
+		for(var/ui in P.dna)
+			if(!dna[ui])
+				uploaded++
+				dna[ui] = 1
+		check_goal()
+		to_chat(user, "<span class='notice'>[uploaded] new datapoints uploaded.</span>")
+	else
+		return ..()
+
+/obj/machinery/dna_vault/proc/upgrade(mob/living/carbon/human/H,upgrade_type)
+	if(!(upgrade_type in power_lottery[H]))
+		return
+	var/datum/species/S = H.dna.species
```

Comment:
Mhm, is there a specific reason we have it set up like that? Would it be problematic to have a new instance for the species of every human?

I have no idea how to make this work otherwise. I'd probably have to re-purpose the DNA vault to give out like free genetics powers.